signer/test/test_user.py

```diff
@@ -0,0 +1,112 @@
+"""
+
+HSM_KEY = SSlibKey(
+    "762cb22caca65de5e9b7b6baecb84ca989d337280ce6914b6440aea95769ad93",
+    "ecdsa",
+    "ecdsa-sha2-nistp256",
+    {
+        "public": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEohqIdE+yTl4OxpX8ZxNUPrg3SL9H\nBDnhZuceKkxy2oMhUOxhWweZeG3bfM1T4ZLnJimC6CAYVU5+F5jZCoftRw==\n-----END PUBLIC KEY-----\n"
+    },
+)
+
+NONCONFIGURED_KEY = SSlibKey(
+    "64eeece964e09c058ef8f9805daca546b01ba4719c80b6fe911b091a7c05124b",
+    "ecdsa",
+    "ecdsa-sha2-nistp256",
+    {
+        "public": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEu+ebm3VUg6U2b0IIeR6NFZU7uxkL\nR1sVLxV8SEW7G+AMXMasEQf5daxfwVMP1kuEkhGs3mBYLkYXlWDh9BNSxg==\n-----END PUBLIC KEY-----\n"
+    },
+)
+
+
+class TestUser(unittest.TestCase):
+    """Test configuration management and signer caching"""
+
+    def test_required(self):
+        with TemporaryDirectory() as tempdir:
+            inifile = os.path.join(tempdir, ".tuf-on-ci-sign.ini")
+            with open(inifile, "w") as f:
+                f.write(REQUIRED)
+
+            user = User(inifile)
+            self.assertEqual(user.name, "@signer")
+            self.assertEqual(user.pykcs11lib, "/usr/lib/x86_64-linux-gnu/libykcs11.so")
+            self.assertEqual(user.push_remote, "origin")
+            self.assertEqual(user.pull_remote, "myremote")
+
+            with open(inifile, "w") as f:
+                f.write(MISSING_NAME)
+            with self.assertRaises(click.ClickException):
+                user = User(inifile)
+
+    def test_signing_keys(self):
+        with TemporaryDirectory() as tempdir:
+            inifile = os.path.join(tempdir, ".tuf-on-ci-sign.ini")
+            with open(inifile, "w") as f:
+                f.write(REQUIRED_AND_SIGNING_KEYS)
+
+            user = User(inifile)
+            # We should get a signer for the configured HSM
+            hsm_signer = user.get_signer(HSM_KEY)
+            self.assertIsInstance(hsm_signer, HSMSigner)
+            self.assertEqual(
+                hsm_signer.token_filter, {"label": "YubiKey PIV #15835999"}
+            )
+            self.assertEqual(
+                hsm_signer.public_key.keyid,
+                "762cb22caca65de5e9b7b6baecb84ca989d337280ce6914b6440aea95769ad93",
+            )
+
+            # If the signing key is not configured, we expect a generic HSM signer
+            other_signer = user.get_signer(NONCONFIGURED_KEY)
+            self.assertIsInstance(other_signer, HSMSigner)
+            self.assertEqual(other_signer.token_filter, {})
+            self.assertEqual(
+                other_signer.public_key.keyid,
+                "64eeece964e09c058ef8f9805daca546b01ba4719c80b6fe911b091a7c05124b",
+            )
+
+            # another lookup should return same instance
+            second_hsm_signer = user.get_signer(HSM_KEY)
+            self.assertIs(hsm_signer, second_hsm_signer)
```

Comment:
Would be nice (and probably easy) to add a "sigstore-oidc" test case. Then you'd have all cases in the `get_signer` switch tested.

Reply:
ah the sigstoresigner can't be tested like this because we currently authenticate at Signer instantiation time: this is arguably a mistake, maybe it should only auth on first sign()